Support SoC and board extensions

tejlmand opened the pull request:
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/69548
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/72374

This PR introduces support for extending boards and SoC with new variants oot.

Scope reduction in this PR compared to #69548:
- Thus it must be possible to describe when a variant want to apply base variant overlay and conf files. 
This feature part will be addressed as follow-up PR.

tejlmand committed:
cmake: scripts: support board extension

Fixes: #69548

Support extending an existing board with new board variants.

This commit introduces the following changes to allow a board to be
extended out-of-tree.

The board yaml schema is extended to support an extend field which
will be used to identify the board to be extended.

A board 'plank' can be extended like this:
> board:
>   extend: plank
>   variants:
>     - name: ext
>       qualifier: soc1

For the rest of the build system this means that there is no longer a
single board directory.
The existing CMake variable BOARD_DIR is kept and reference the
directory which defines the board.
A new CMake variable BOARD_DIRECTORIES provides a list of all
directories which defines board targets for the board.
This means the directory which defines the board as well as all
directories that extends the board.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: cleanup hwm_v2.cmake module code

Cleanup the Kconfig generating code in hwm_v2.cmake by moving common
logic inside the kconfig_gen() helper function.

This prepares the code for board extension feature.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
tests: add twister tests for CMake board and SoC extension

This commit adds new tests for verifying the functionality of the board
and SoC extension feature.

It does so by defining:
- A new CPU cluster on an existing SoC
- Two new board variants on top of an existing board

The new board variants are defined on top of the existing `native_sim`
board, so that the following native_sim board targets are available for
the test.
Existing board targets:
- native_sim/native
- native_sim/native/64

Extended board targets:
- native_sim/native/one
- native_sim/native/64/two

The new CPU cluster is defined for the existing `an521` SoC.
Existing CPU Clusters on an521:
- cpu0
- cpu1

New CPU Cluster:
- cputest

For SoC tests the mps2 board is used.
This means that for testing, the following board targets using the an521
SoC are:
- mps2/an521/cpu0
- mps2/an521/cpu1
- mps2/an521/cputest

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: scripts: support SoC extension

Fixes: #72374

Support extending an existing SoC with new CPU clusters.

This commit introduces the following changes to allow an SoC to be
extended out-of-tree.

The SoC yaml schema is extended to support an extend field which
will be used to identify the SoC to be extended with extra CPU clusters.

A SoC 'a_soc' can be extended like this:
> socs:
>   extend: a_soc
>   cpuclusters:
>     - name: extra_core

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nashif commented:
nice. would be great if this feature is presented in the arch wg to get more feedback and input.

tejlmand committed:
doc: extend board porting guide with new board extend feature

Extend the board porting guide with documentation on the new board
extend feature.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

henrikbrixandersen commented:
@PerMac Couldn't this be used for addressing https://github.com/zephyrproject-rtos/zephyr/issues/73421 as well?

PerMac commented:
> @PerMac Couldn't this be used for addressing #73421 as well?

I will sync with @tejlmand to get a better understanding.

57300 commented on cmake/modules/dts.cmake:
```diff
@@ -86,7 +86,7 @@ find_package(Dtc 1.4.6)
 # - DTS_SOURCE: the devicetree source file to use may be pre-set
 #   with this variable; otherwise, it defaults to
-#   ${BOARD_DIR}/${BOARD}.dts
+#   ${BOARD_DIRECTORIES}/${BOARD}.dts
```
nit: `${BOARD}.DTS` is a HWMv1 leftover. Maybe now is a good time to update it?

tejlmand replied:
Fixed

57300 commented on cmake/modules/kconfig.cmake:
```diff
@@ -157,7 +160,7 @@ set(COMMON_KCONFIG_ENV_SETTINGS
   CONFIG_=${KCONFIG_NAMESPACE}_
   KCONFIG_CONFIG=${DOTCONFIG}
-  BOARD_DIR=${BOARD_DIR}
+  BOARD_DIR=${KCONFIG_BOARD_DIR}
```
Seems like a breaking change, depending on how `$(BOARD_DIR)` is used out of tree. I'd suggest adding `KCONFIG_BOARD_DIR` as a separate Kconfig macro.

I suppose that the Kconfig sources should be updated to use `$(KCONFIG_BINARY_DIR)/boards` at some point, but right now, HWMv1 is getting in the way.

tejlmand replied:
> Seems like a breaking change

No, for HWMv1 then existing behavior for BOARD_DIR is the same:
https://github.com/zephyrproject-rtos/zephyr/blob/d0dbcc41e1af58a7dae70b6a5d571e2d1bacda38/cmake/modules/kconfig.cmake#L27-L28

For HWMv2 this is not a breaking change because no Zephyr releases until today has included HWMv2.
Upcoming Zephyr v3.7 will be first to include HWMv2, and therefore we can still make changes to how this works for that release.

57300 replied:
I guess it's fair to point out that the meaning of `BOARD` also changed with HWMv2, but what hasn't changed is the equality of CMake variable and Kconfig macro. The fact that they're different for `BOARD_DIR` might be a head-scratcher.

tejlmand replied:
> The fact that they're different for BOARD_DIR might be a head-scratcher.

That's a very valid point.
Will address.

57300 commented on doc/_extensions/zephyr/kconfig/__init__.py:
```diff
@@ -91,7 +91,7 @@ def kconfig_load(app: Sphinx) -> Tuple[kconfiglib.Kconfig, Dict[str, str]]:
         v2_systems = list_hardware.find_v2_systems(root_args)
 
-        soc_folders = {soc.folder for soc in v2_systems.get_socs()}
+        soc_folders = {soc.folder[0] for soc in v2_systems.get_socs()}
```
If board/SoC extensions will be allowed in-tree, will this have to be updated to include all folders?
```suggestion
        soc_folders = {folder for soc in v2_systems.get_socs() for folder in soc.folder}
```
Same with the occurrences of `board.dir[0]`.

tejlmand replied:
that depends on how we want to organize the docs in such case.

Will we want docs describing the in-tree soc / board to be located in a single folder, or do we want to allow documentation for a given soc / board to be placed at multiple locations.

I would like to leave such decision to doc maintainers and not as part of this PR.
@kartben thoughts ?

57300 replied:
OK, I would agree that the docs can be handled later, same with compliance.

57300 commented on scripts/ci/check_compliance.py:
```diff
@@ -844,6 +845,9 @@ def check_no_undef_outside_kconfig(self, kconf):
                               # Zephyr toolchain variant and therefore not
                               # visible to compliance.
         "BOARD_", # Used as regex in scripts/utils/board_v1_to_v2.py
+        "BOARD_MPS2_AN521_CPUTEST", # Used for board and SoC extension feature
+        "BOARD_NATIVE_SIM_NATIVE_64_TWO", # Used for board and SoC extension feature
+        "BOARD_NATIVE_SIM_NATIVE_ONE", # Used for board and SoC extension feature
```
nit: Specifically, they're used for board and SoC extension feature _tests_.

tejlmand replied:
fixed

57300 commented on scripts/list_boards.py:
```diff
@@ -91,7 +91,7 @@ def from_soc(soc, variants):
 class Board:
     name: str
-    dir: Path
+    dir: List[Path]
```
nit: Maybe `directories`, both to help clarify that this is a list and to align with CMake.

tejlmand replied:
Good suggestion, applied.

57300 commented on scripts/list_boards.py:
```diff
@@ -102,6 +102,41 @@ class Board:
+            else:
+                node = Variant(None)
+
+        if node == Variant(None):
```
Should also handle unknown SoC for completeness.
```suggestion
        if node == in (Soc(None), Variant(None)):
```

tejlmand replied:
fixed

57300 commented on scripts/list_boards.py:
```diff
@@ -219,9 +267,9 @@ def load_v2_boards(board_name, board_yml, systems):
                     for s in board.get('socs', {})]
 
-            board = Board(
+            boards.update({board['name']: Board(
```
Isn't this the same as:
```suggestion
            boards[board['name']] = Board(
```

tejlmand replied:
yes, guess I just was having `update` from other places.
Do you prefer a change, or is it simply because the other is shorter ?

57300 replied:
I think this change would be cleaner, especially compared to having `)})` at the bottom of the statement.

tejlmand replied:
fair point, fixed

57300 commented on scripts/list_boards.py:
```diff
@@ -251,14 +310,21 @@ def find_v2_boards(args):
-        b = load_v2_boards(args.board, board_yml, systems)
-        boards.extend(b)
+        b, e = load_v2_boards(args.board, board_yml, systems)
+        boards.update(b)
```
Should we make sure that no two boards can share the same `name`, both here and in `load_v2_boards()`?
Especially considering that this is no longer checked in CMake.

tejlmand replied:
good point.
A board can be extended but two boards should not both create conflicting boards.

tejlmand replied:
Fixed, if duplicate boards are now discovered then list_boards will print an error:
https://github.com/zephyrproject-rtos/zephyr/blob/41561f164a8f7bc1e697fd1f7942700dd4b54854/scripts/list_boards.py#L327-L330

57300 commented on tests/cmake/hwm/board_extend/oot_root/boards/arm/mps2/mps2_an521-common.dtsi:
```diff
@@ -0,0 +1,191 @@
+timer0: timer@0 {
+	compatible = "arm,cmsdk-timer";
+	reg = <0x0 0x1000>;
+	interrupts = <3 3>;
```
Wouldn't it be better to include the original board DTS, instead of copying it into the test?

tejlmand replied:
a copy was chosen as we just need a device tree where we can verify that it's being correctly extended.
We have no intention of tracking the real board, and only reason to have the real board in this case is to ensure the extension mechanism works.

57300 replied:
> We have no intention of tracking the real board

If that's the case, have you considered using something closer to blank/dummy files? I don't know if the dts contents here are essential to pass these tests.

tejlmand replied:
I did, but this is where I took the short route of a copy and instead focus on the board extend functionality itself.

Having the dts as close to the board it extends is probably more likely than an blank / dummy dts file.
Although you can say that for a real extend, a user would most likely try to include a dtsi file if possible, which is close to your first comment 😉

57300 reviewed (commented):
Tested this a bunch over the past few weeks and it works as expected. Thanks!
Just a couple of comments on the implementation.

nordicjm commented on scripts/list_boards.py:
```diff
@@ -198,6 +234,18 @@ def load_v2_boards(board_name, board_yml, systems):
+                sys.exit(f'ERROR: Malformed "board" section in file: {board_yml.as_posix()}\n'
+                         f'{mutual_exclusive} are mutual exclusive at this level.')
+
+            # This is a extending an exsisting board, place in array to allow later processing.
```
*existing

tejlmand replied:
fixed

nordicjm commented on cmake/modules/hwm_v2.cmake:
```diff
@@ -73,7 +73,7 @@ while(TRUE)
     string(TOUPPER "${ARCH_V2_NAME}" ARCH_V2_NAME_UPPER)
     set(ARCH_V2_${ARCH_V2_NAME_UPPER}_DIR ${ARCH_V2_DIR})
```
this comment has "a board cos"
*soc

tejlmand replied:
Rephrased.

nordicjm reviewed (commented):
an extended board cannot change the west flash runner config it seems because it looks for that in `scripts/west_commands/run_common.py`:
```
            if cache['BOARD_DIR'] not in processed_boards and 'SOC_FULL_DIR' in cache:
                soc_yaml_file = Path(cache['SOC_FULL_DIR']) / 'soc.yml'
                board_yaml_file = Path(cache['BOARD_DIR']) / 'board.yml'
                group_type = 'boards'
```
Which does not support using the extended board config but seems like something that should be used also (if the file exists in the extended board folder, if not it uses the original board file if it exists, otherwise the soc file)

nordicjm reviewed (commented):
Can also seemingly extend a board twice, which is invalid, by duplicating the extended board, the output gives this:
```
CMake Error at /tmp/aa/zephyr/cmake/modules/boards.cmake:304 (message):
  Board qualifiers `` for board `nrf52840dk` not found.  Please specify a
  valid board target.

  Valid board targets for nrf52840dk are:
  nrf52840dk/nrf52840
  nrf52840dk/nrf52840/ext
  nrf52840dk/nrf52840/ext
  nrf52840dk/nrf52811
```
And you can build for `nrf52840dk/nrf52840/ext`

```
-- Found BOARD.dts: /tmp/aa/modules/hal/nordic/boards/bob2/nrf52840dk_nrf52840_ext.dts
Loaded configuration '/tmp/aa/modules/hal/nordic/boards/bob/nrf52840dk_nrf52840_ext_defconfig'
Merged configuration '/tmp/aa/modules/hal/nordic/boards/bob2/nrf52840dk_nrf52840_ext_defconfig'
```

tejlmand commented:
> Can also seemingly extend a board twice, which is invalid, by duplicating the extended board, the output gives this:

Fixed, an error will now be printed:
https://github.com/zephyrproject-rtos/zephyr/blob/41561f164a8f7bc1e697fd1f7942700dd4b54854/scripts/list_boards.py#L294-L297

57300 commented on scripts/list_boards.py:
```diff
@@ -232,8 +279,23 @@ def load_v2_boards(board_name, board_yml, systems):
+
+        for v in e.get('variants', []):
+            node = board.from_qualifier(v['qualifier'])
+            if str(v['qualifier'] + '/' + v['name']) in board_v2_qualifiers(board):
+                board_yml = e['dir'] / BOARD_YML
+                sys.exit(f'ERROR: Variant: {v["name"]}, defined multiple times for board: '
+                         f'{board.name}.\nLast defined in {board_yml}')
```
This check should be made more general. It catches duplicate qualifiers added by extensions, but doesn't catch duplicate qualifiers in the main board definition:
```yaml
board:
  name: nrf54h20dk
  socs:
  - name: nrf54h20
    variants:
    - name: xip
      cpucluster: cpuppr
    - name: xip
      cpucluster: cpuppr
```

tejlmand replied:
> but doesn't catch duplicate qualifiers in the main board definition:

True, and I will take a look.
But in practice, a duplicate entry in the main board definition is not an issue in practice because the normalized board target will still pick up the same files from the same board folder, in your example `<main_board_dir>/nrf54h20dk_nrf54h20_cpuppr_xip_defconfig`.

The real issue with duplicate / colliding qualifiers from main board vs extended board is related to ambiguity on where a board variant is really defined, meaning from where will it be picked up.

tejlmand replied:
Added a qualifier post check when creating a new Board instance.

57300 commented on scripts/list_boards.py:
```diff
@@ -363,7 +432,7 @@ def dump_boards(args):
                 info = args.cmakeformat.format(
                     NAME='NAME;' + board.name,
-                    DIR='DIR;' + str(board.dir.as_posix()),
+                    DIR='DIR;' + str(board.directories.as_posix()),
```
`List[Path]` type should be respected:
```suggestion
                    DIR='DIR;' + str(board.directories[0].as_posix()),
```

And `find_arch2board_set_in()` should be updated:
```diff
-ret[arch].add(Board(board_name, maybe_board, 'v1', arch=arch))
+ret[arch].add(Board(board_name, [maybe_board], 'v1', arch=arch))
```

tejlmand replied:
True, a bit of misuse.

The reason is that the existing board v1 code relies on the Board dataclass to be hashable.
https://github.com/zephyrproject-rtos/zephyr/blob/d0a2451b6de89ec9532653674a8d614a1eac25d6/scripts/list_boards.py#L161
We want the ability to extend the board in v2, but that also means we cannot just make it frozen.

Although a bit hack'ish, then considering that we want to get rid of hwmv1, then I felt keeping the current use could be accepted, but I agree a comment should be added in this case.

57300 replied:
> The reason is that the existing board v1 code relies on the Board dataclass to be hashable.

Good point, I missed that. Then, instead of touching this code, I think it would make sense to change the type annotation to `Union[Path, List[Path]]` and put a comment there.

tejlmand replied:
Fixed.

57300 commented on scripts/ci/test_plan.py:
```diff
@@ -242,7 +242,7 @@ def find_boards(self):
         for changed in changed_boards:
             for board in known_boards:
```
Remember that `known_boards` is now a dict.

nordicjm reviewed (changes requested):
Not addressed:

an extended board cannot change the west flash runner config it seems because it looks for that in `scripts/west_commands/run_common.py`:
```
            if cache['BOARD_DIR'] not in processed_boards and 'SOC_FULL_DIR' in cache:
                soc_yaml_file = Path(cache['SOC_FULL_DIR']) / 'soc.yml'
                board_yaml_file = Path(cache['BOARD_DIR']) / 'board.yml'
                group_type = 'boards'
```
Which does not support using the extended board config but seems like something that should be used also (if the file exists in the extended board folder, if not it uses the original board file if it exists, otherwise the soc file)

tejlmand commented:
> an extended board cannot change the west flash runner config it seems because it looks for that in scripts/west_commands/run_common.py:

Correct, this use-case is not described in https://github.com/zephyrproject-rtos/zephyr/issues/72374 or https://github.com/zephyrproject-rtos/zephyr/issues/69548 and therefore not part of this PR.

> Which does not support using the extended board config but seems like something that should be used also (if the file exists in the extended board folder, if not it uses the original board file if it exists, otherwise the soc file)

This would be a too simplified solution, as it would expect that a board can only be extended once.
If we want to allow a user to overwrite or extend the runners oot, then we should actually have a dedicated issue opened, so that it can be properly discussed how such extension is expected to work.

Just following the proposal in the comment will not work in case a board is extended twice.
Instead it should be considered if an extended board should be able to extend the runners section with additional information, but that would require merging of runner information from multiple sources.

Or, if we decide to just load the runners information once, then we would need to discuss and decide who takes precendence when a board is extended multiple times.

Or should it be so that the extended `runners` section is only loaded if the extended board is used.
Note, such solution would require that we keep track of the location where the extended board is defined.

Due to above, I think extension of runners deserves its own issue and PR, so that a proper solution can be made, and therefore is outside the scope of this PR.

tejlmand commented:
> It needs to be part of the base PR

I disagree.
The flash runner feature itself was made after the HWMv2 itself, and similar can extension of flash runners be made after extension of boards. Flash runners are a west feature aimed at flashing, whereas board and soc extensions are aimed at building images.

> an issue can be opened specifically for it for discussion purposes but it must be completely implemented in the original feature PR.

An issue should be opened for dedicated discussion.
Discussing the flash runner extension in this PR just makes noise.

nordicjm committed:
scripts: west_commands: run: Add priority to run once config

Adds an optional priority parameter to the flash runner run once
configuration which allows for deciding upon which file should
ultimately be used

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
scripts: list_hardware: Do not check qualifiers for run once config

Removes validating the qualifiers for flash run once configuration
as files may be present that contain information for qualifiers
that are not present in a single repository but are spaced out in
other repositories, or might be optional

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

PerMac reviewed (changes requested):
breaks testing on v1 boards 
https://github.com/zephyrproject-rtos/zephyr/pull/76055

57300 commented:
> breaks testing on v1 boards

@PerMac Thanks for catching this. I can push a fix here while @tejlmand is on vacation.
I'm surprised to find multiple instances of the missing `Board.dir -> Board.directories` rename.

PerMac commented:
@57300 maybe then it would be safer to "revert" change from dir to directories in this PR if there can be more misaligned places?

57300 commented:
> @57300 maybe then it would be safer to "revert" change from dir to directories in this PR if there can be more misaligned places?

The rename is still useful IMO (it was my suggestion) as it matches the `BOARD_DIRECTORIES` CMake variable.
But `BOARD_DIR` is still defined in CMake as well, so I'm thinking of adding this to the `Board` class:
```py
@property
def dir(self):
    return self.directories[0]
```
Then the changes in other scripts can be reverted.

57300 committed:
scripts: Undo Board.{dir -> directories[0]} renames

`Board.dir` can be used in place of `Board.directories[0]` for brevity.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
scripts: list_boards: Reintroduce Board.dir as @property

In the `Board` class, the `dir` member was renamed to `directories`,
both to indicate that it is now a list (in HWMv2 with board extensions)
and to reflect the addition of the `BOARD_DIRECTORIES` CMake variable.
Considering that the build system also keeps the `BOARD_DIR` variable,
and for the sake of backwards compatibility and brevity, it should be
useful to retain `Board.dir` in Python as well, symmetrically.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

teburd commented:
I don't see any docs in the PR explaining how this works or what I'd need to do to try it out. How would I go about trying this out?

kartben commented on doc/_scripts/gen_boards_catalog.py:
```diff
@@ -70,15 +70,15 @@ def get_catalog():
         board_roots=module_settings["board_root"],
         soc_roots=module_settings["soc_root"],
-        board_dir=ZEPHYR_BASE / "boards",
+        board_dir=[ZEPHYR_BASE / "boards"],
```
```suggestion
        board_dir=[],
```

this seems to do the trick. No idea why. Also please consider renaming to `board_dirs` if this is now an array

tejlmand replied:
> it looks like I may have misunderstood what "board_dir" is for

board dir is intended to indicate locations where to look for boards.
This is useful during incremental builds where the board dirs are cached by CMake and thus we can pass the exact directories to `list_boards.py` and thereby avoid searching all folders for a given board.

The reason we still invoke list_boards even when we know the location is to be able to both detect renames, folder movement, changes in board variants, etc., and there the list_board.py has all the knowledge.

Will change to `[]`, as that seems to be what you actually want in this case.

Regarding rename to `board_dirs` then you have point when the list_boards is imported into another Python, but as `--board-dir` is actually a command line argument, which can be given multiple times, like `list_boards.py --board-dir=<path>/foo-board --board-dir=<path>/bar-board --board-dir=...`, then I actually think the use of singular in the name in this case is correct.

tejlmand replied:
changed to `[]`.

carlescufi reviewed (dismissed):
LGTM, late in the 4.0.0 cycle so up to the release team whether to include it.

tejlmand commented:
> This has been open for months without activity at all and without further reviews, so it was not clear where it was going.

@nashif  the only thing which has happened is a rebase, there has been no functional changes because it has been ready.
The issues seen in CI was minor and due to some `list_boards` calls in docs which was added between last rebase and the newest rebase.

>
 We need more reviews here and testing before it can be merged and for 4.0, this is just too late as the PR basically changed status without any warning at the very last minute.
> 

which status change are you referring to ?

nordicjm commented:
I have checked this out this morning by adding an extension to nrf5340dk then checking that the extended board was used and that the flash runner configuration from the extended board is used

nashif commented:
> which status change are you referring to ?

status change from what seemed to be a completely abandoned PR with pending change requests (stale bot marked it stale just 5 days ago), to approved at the last minute before feature freeze. This is basic infrastructure changes and need further review/testing by others.

kartben commented:
FWIW I agree with @nashif and was about to comment something similar.
While documentation for this *could* be added later, the fact that @teburd's [comment](https://github.com/zephyrproject-rtos/zephyr/pull/72857#issuecomment-2299410295) hasn't been addressed is IMO a sign that this might not be ready just yet.

To put it differently, for this to be reviewed effectively, it needs to be something that can be reviewed from a *user* perspective, not from a Python/Kconfig/Cmake code perspective. I, for one, would like to provide feedback on the overall user experience for this new (very much welcome, don't get me wrong) feature, and I simply wouldn't know where to start

tejlmand commented:
> While documentation for this _could_ be added later, the fact that @teburd's [comment](https://github.com/zephyrproject-rtos/zephyr/pull/72857#issuecomment-2299410295) hasn't been addressed is IMO a sign that this might not be ready just yet.

sorry, I had missed that comment.
Docs are here: https://github.com/zephyrproject-rtos/zephyr/pull/73409

nashif commented:
> sorry, I had missed that comment.

I do not blame you, all I can see in this PR is:

![image](https://github.com/user-attachments/assets/4045aeda-0035-4926-a1dd-0ef8c5132cc4)

carlescufi commented:
> > sorry, I had missed that comment.
> 
> I do not blame you, all I can see in this PR is:

@nashif Yes, we are addressing this as we speak, you will no longer see those in upstream PRs. Sorry for the noise.

tejlmand commented:
> @nashif Yes, we are addressing this as we speak, you will no longer see those messages in upstream PRs. Sorry for the noise.

at least this long and horrible list of noise :see_no_evil:  shows that the feature is cherry-picked into NCS because it's used downstream. Thus it's actively being tested, besides the twister tests included in the PR.

tejlmand commented:
> pending change requests 

those change requests was actually addressed months ago, but true, I didn't followup with reviewers to get final approvals.

nashif commented:
> at least this long and horrible list of noise 🙈 shows that the feature is cherry-picked into NCS because it's used downstream. Thus it's actively being tested, besides the twister tests included in the PR.

i have no idea what that is and should not be relying on this data to make judgment about the state of a PR and its level of testing and maturity. I am sure it works for you, but this a project feature and not a nordic only feature and a core one that requires more reviews and testing.

nashif reviewed (changes requested):
> sorry, I had missed that comment.
> Docs are here: #73409

why is this in a different PR? This should come as part of the same PR and go in together.

tejlmand commented:
> but this a project feature and not a nordic only feature and a core one that requires more reviews and testing.

and it has been developed as a project feature, not a Nordic feature.
But project features are developed because they are useful, and being useful sometimes also means useful to Nordic.
And in this case cherry-picked into Nordics fork. 

Nordic doesn't require this PR for 4.0.0, because we already have it cherry-picked, but the feature can be very useful to other users / companies, and for that I think it is beneficial if it is part of 4.0.0.

As side-note what testing are reviewers doing on build system PRs ?
I rarely find people who actually tests build system features in depth before they are merged.

Even good reviews  on the build system itself is hard to get :disappointed: .

If you mean more testing on main, before being part of a release, then that still requires someone downstream of Zephyr repo to use this feature, because Zephyr main itself will not use this feature.
The feature allows Zephyr downstream users to extend upstream boards.
That's one reason that I think it's worth mentioning that it is being used by Nordic, because it means the use is actually verified.

> why is this in a different PR? This should come as part of the same PR and go in together.

Simply because I often experience that you get approvals on the implementation, just to get change request on docs afterwards, resulting in having to again ping reviewers to get re-approvals. Or vice-versa.

Having two PRs avoids approvals for the implementation to be cleared when having to push change requests on the doc set.

carlescufi commented:
@kartben  regardless of when this gets merged, are you happy with the doc?

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -912,9 +912,69 @@ There are some extra things you'll need to do:
 Board extensions
 ****************
 
+The board hardware model in Zephyr allows you to extend an existing board with
+new board variants. Such board extensions can be done in your custom repository
+and thus outside of the Zephyr repository.
+Zephyr's board hardware model allows you to extend existing boards with new
+variants. Such board extensions can be done in your custom repository and thus
+outside of the Zephyr repository.
```
repeated. I tend to like second paragraph's formulation better, fwiw
also would recommend skipping a line so that "Extending an existing board with [...]" is another paragraph

tejlmand replied:
of course you like the second paragraph best, cause you wrote it yourself :wink: 
I happened to just apply your suggestion in this comment https://github.com/zephyrproject-rtos/zephyr/pull/73409#discussion_r1636802273 using the `commit suggestion` feature of github.

Guess I should have noticed that you were actually not removing the old lines in that suggestio :laughing: 

Repeated paragraph removed.

kartben reviewed (changes requested):
>@kartben regardless of when this gets merged, are you happy with the doc?

@carles it looks ok - one repeated paragraph tho

henrikbrixandersen reviewed (approved):
I often use the existing board extension mechanism, but it is very limited. This will solve my current issues with that, and open up for much smoother (customised) hardware-in-the-loop testing here. Thanks.

kartben reviewed (approved):
+1 for docs

nashif commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -912,9 +912,67 @@ There are some extra things you'll need to do:
+   ├── Kconfig
+   ├── Kconfig.plank
+   ├── Kconfig.defconfig
+   └── plank_<new-qualifiers>.yaml
```
had to scratch my head a bit here trying to understand what the new_qualifier would end up like by looking at the tests. It will help alot if you provide an example here showing how you get to the new qualifier by extending a board.

nashif replied:
maybe:

As an example, when extending the native_sim board using the following:

```
board:
  extend: native_sim
  variants:
    - name: one
      qualifier: native
    - name: two
      qualifier: native/64
```


the new board targets will end up as `native_sim/native/one` and `native_sim/native/64/two`. The DTS files for the extended boards will be `native_sim_native_64_two.dts` and `native_sim_native_one.dts`.

nashif commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -912,9 +912,67 @@ There are some extra things you'll need to do:
 Board extensions
 ****************
 
+The board hardware model in Zephyr allows you to extend an existing board with
```
The PR also does add SoC extension out of tree, where is this documented?

nashif reviewed (approved):
please submit followup PRs expanding the docs:
- add an example with real names o illustrate how a new board extension will look like when it comes to names
- add docs about SoC extensions
